SSA/Projects/InstCombine/scripts/mlir-tool.py

```diff
@@ -0,0 +1,183 @@
+#!/usr/bin/env python3
+from xdsl.dialects.builtin import ModuleOp
+from xdsl.dialects.llvm import LLVM
+from xdsl.utils.exceptions import ParseError
+from xdsl.context import MLContext
+from xdsl.dialects import get_all_dialects
+from xdsl.dialects.llvm import FuncOp
+from xdsl.parser import Parser
+from xdsl.printer import Printer
+from xdsl.dialects.builtin import (
+    Builtin,
+    IndexType,
+    IntegerAttr,
+    IntegerType,
+    ModuleOp,
+    StringAttr,
+    i32,
+    i64,
+)
+import os
+import io
+from pathlib import Path
+from xdsl.printer import Printer
+
+# Initialize the MLIR context and register the LLVM dialect
+ctx = MLContext(allow_unregistered=True)
+ctx.load_dialect(LLVM)
+ctx.load_dialect(Builtin)
+
+directory = os.fsencode("../vcombined-mlir")
+
+
+allowed_names = {
+    "llvm.return",
+    "llvm.mul",
+    "llvm.add",
+    "llvm.sub",
+    "llvm.shl",
+    "llvm.and",
+    "llvm.or",
+    "llvm.xor",
+    "llvm.mlir.constant",
+    "llvm.lshr",
+    "llvm.ashr",
+    "llvm.urem",
+    "llvm.srem",
+    "llvm.add",
+    "llvm.mul",
+    "llvm.sub",
+    "llvm.sdiv",
+}
+allowed_unregistered = set()  # {
+
+
+# "llvm.icmp"
+# }
+def allowed(op):
+    return (
+        hasattr(op, "sym_name")
+        or (
+            op.name == "builtin.unregistered"
+            and op.op_name.data in allowed_unregistered
+        )
+        or (op.name in allowed_names)
+    )
+
+
+def show(block):
+    output = io.StringIO()
+    p = Printer(stream=output)
+    block.print(p)
+    contents = output.getvalue()
+    output.close()
+    return contents
+
+
+def showr(region):
+    output = io.StringIO()
+    p = Printer(stream=output)
+    p.print_region(region)
+    contents = output.getvalue()
+    output.close()
+    return contents
+
+
+def size(func):
+    return sum(1 for _ in func.walk())
+
+
+def read_file(file_name):
+    with open(file_name, "r") as f:
+        return f.read()
+
+
+for file in os.listdir(directory):
+    filename = os.fsdecode(file)
+    print(filename)
+    stem = "g" + filename.split(".")[0].replace("-", "h")
+    output = ""
+    data1 = read_file(os.path.join("../vcombined-mlir"))
+    data2 = read_file(os.path.join("../vbefore-mlir"))
+    parser1 = Parser(ctx, data1)
+    try:
+        module1 = parser1.parse_module()
+        parser2 = Parser(ctx, data2)
+        module2 = parser2.parse_module()
+        funcs = [
+            func
+            for func in module1.walk()
+            if isinstance(func, FuncOp)
+            and all(allowed(o) for o in func.walk())
+            and size(func) > 1
+        ]
+        for func in funcs:
+            try:
+                other = next(
+                    f
+                    for f in module2.walk()
+                    if isinstance(f, FuncOp) and f.sym_name == func.sym_name
+                )
+                if all(allowed(o) for o in other.walk()):
```

Comment:
I would recommend caching these and accessing in O(1) outside the loop:
```suggestion
funcs2 = {
    f.sym_name: f
    for f in module2.walk()
    if isinstance(f, FuncOp)
    if f.sym_name == func.sym_name
    if all(allowed(o) for o in other.walk())
}
```